### PR TITLE
feature/SCKT-1_init_project_supaapps_api_kit_client

### PR DESCRIPTION
**Pull request by MostafaFawzy7**

### CODE REVIEW

File: src/ApiKitClient.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+      this.unauthorizationCallback = unauthorizationCallback;
+
+      this.instance.interceptors.response.use((response: AxiosResponse) => response, (error: AxiosError) => {
+        if (error.response?.status === 401 && this.unauthorizationCallback) {
+          this.unauthorizationCallback();
+        }
+        return Promise.reject(error);
+      });
+    }
```
Comment (Abdallah-Abdelkhalek) [inline src/ApiKitClient.ts]: please explain why it always returns Promise.reject(error)

Comment (MostafaFawzy7) [reply src/ApiKitClient.ts]: Not always, when there is be error, will return an error in a shortcut way!
Like the `catch(err)` statement...

Comment (Abdallah-Abdelkhalek) [reply src/ApiKitClient.ts]: I mean this functions will still return `Promise.reject(error);`

Comment (MostafaFawzy7) [reply src/ApiKitClient.ts]: You can inside the consumer app adding the promise rejection chain method 
```js
.catch(err => {
   // Do logic of rejection here
})
```

**Comment by SirNarsh** [comment]
missing check if not intialized

File: src/ApiKitClient.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  public static async get<T>(endpoint: string, params?: URLSearchParams): Promise<AxiosResponse<PaginatedResponse<T[]>>> {
+    return this.instance.get<PaginatedResponse<T[]>>(endpoint, { params });
```
Comment (SirNarsh) [inline src/ApiKitClient.ts]: what if endpoint is not paginated

**Comment by SirNarsh** [comment]
authtoken as an async method, check supaapps-auth

**Comment by MostafaFawzy7** [comment]
@SirNarsh All comments addressed!

File: src/ApiKitClient.ts [hunk 3]
```diff
@@ -2,21 +2,31 @@ import axios, { AxiosInstance, AxiosResponse, AxiosError } from 'axios';
 import { PaginatedResponse } from './types';
 
 type UnauthorizationCallback = () => void;
+type AuthTokenCallback = () => Promise<string>;
 
 export class ApiKitClient {
-  private static instance: AxiosInstance | null = null;
+  private static instance: AxiosInstance;
   private static unauthorizationCallback?: UnauthorizationCallback;
+  private static authTokenCallback?: AuthTokenCallback;
```
Comment (Abdallah-Abdelkhalek) [inline src/ApiKitClient.ts]: authTokenCallback optional?

Comment (MostafaFawzy7) [reply src/ApiKitClient.ts]: Missed this by accident!
Made both callbacks not optional...

File: README.md [hunk 4]
```diff
@@ -1 +1,125 @@
...
+
+const BASE_URL = 'https://api.example.com'
+
+const AUTH_TOKEN = 'your_auth_token'
+
+const unauthorizationCallback = () => {
+  // Implement redirection to login page here
+};
+
+ApiKitClient.initialize(BASE_URL, AUTH_TOKEN, unauthorizationCallback);
```
Comment (SirNarsh) [inline README.md]: AUTH_TOKEN string is invalid anymore

File: src/ApiKitClient.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  public static async patch<T>(endpoint: string, data: T): Promise<AxiosResponse<T>> {
+    this.checkInitialization();
+    return this.instance!.patch<T>(endpoint, data);
+  }
+
+  public static async delete<T>(endpoint: string): Promise<AxiosResponse<T>> {
+    this.checkInitialization();
+    return this.instance!.delete<T>(endpoint);
```
Comment (SirNarsh) [inline src/ApiKitClient.ts:76]: POST & DELETE topic, I remember we have an old apache php-fpm bug where we always need to send delete as `_METHOD` query

**Comment by SirNarsh** [comment]
let's also add another file that at least `npm install` , `npm build` to check that the code can build. also we need to add testing later on

**Comment by MostafaFawzy7** [comment]
@SirNarsh All comments are addressed!

File: src/ApiKitClient.ts (hunk 5 above)
Comment (MostafaFawzy7) [reply src/ApiKitClient.ts:76]: Postponed this for later iteration as we agreed!

File: README.md (hunk 4 above)
Comment (MostafaFawzy7) [reply README.md]: ✅

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+name: Build
+
+on:
+  push:
+    branches:
+        - main
+        - feature/SCKT-1_init_project_supaapps_api_kit_client
```
Comment (SirNarsh) [inline .github/workflows/build.yml]: all branches should test & build not only main & this branch so we can always test PR
Only publishing is exclusive to tags

Comment (MostafaFawzy7) [reply .github/workflows/build.yml]: Done ✅

**Review by SirNarsh — APPROVED** [review]

**Pull request merged**